fpga_interchange/fasm_generators/xc7.py

```diff
@@ -0,0 +1,92 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2020  The SymbiFlow Authors.
+#
+# Use of this source code is governed by a ISC-style
+# license that can be found in the LICENSE file or at
+# https://opensource.org/licenses/ISC
+#
+# SPDX-License-Identifier: ISC
+from fpga_interchange.fasm_generators.generic import FasmGenerator
+from fpga_interchange.route_stitching import flatten_segments
+from fpga_interchange.physical_netlist import PhysicalPip
+
+
+class XC7FasmGenerator(FasmGenerator):
+    def handle_ios(self):
+        """
+        This function is specialized to add FASM features for the IO buffers
+        in the 7-Series database format.
+        """
+
+        # FIXME: Need to make this dynamic, and find a suitable way to add FASM annotations to the device resources.
+        #        In addition, a reformat of the database might be required to have an easier handling of these
+        #        features.
+        allowed_io_types = {
+            "OBUF": [
+                "LVCMOS12_LVCMOS15_LVCMOS18_LVCMOS25_LVCMOS33_LVTTL_SSTL135_SSTL15.SLEW.SLOW",
+                "LVCMOS33_LVTTL.DRIVE.I12_I16", "PULLTYPE.NONE"
+            ],
+            "IBUF": [
+                "LVCMOS12_LVCMOS15_LVCMOS18_LVCMOS25_LVCMOS33_LVTTL.SLEW.FAST",
+                "LVCMOS12_LVCMOS15_LVCMOS18_LVCMOS25_LVCMOS33_LVDS_25_LVTTL_SSTL135_SSTL15_TMDS_33.IN_ONLY",
+                "LVCMOS25_LVCMOS33_LVTTL.IN", "PULLTYPE.NONE"
+            ]
+        }
+
+        iob_sites = ["IOB_Y1", "IOB_Y0"]
+
+        for cell_instance, cell_data in self.physical_cells_instances.items():
+            if cell_data.cell_type not in allowed_io_types:
+                continue
+
+            iob_site_idx = cell_data.sites_in_tile.index(cell_data.site_name)
+
+            iob_site = iob_sites[
+                iob_site_idx] if "SING" not in cell_data.tile_name else "IOB_Y0"
+
+            for feature in allowed_io_types[cell_data.cell_type]:
+                self.cells_features.append("{}.{}.{}".format(
+                    cell_data.tile_name, iob_site, feature))
+
+    def handle_site_thru(self, site_thru_pips):
```

Comment:
I think handling of LUT route-through PIPs is going to be needed here too

Reply:
Yep, I modified the FF test in nextpnr to get to use FFs in SLICEs (happens often in small tests that the ILOGIC/OLOGIC FFs are used instead) and this indeed showed the need of handling these PIPs. I might generalize a bit more how the LUT truth table is generated to better handle this case.

I guess it will be critical to correctly handle cases where, for instance, the LUT6 is used as a route-thru and the LUT5 is used as a normal LUT (I guess this is possible right?) therefore requiring the final FASM feature to enable both of them.

Reply:
> the LUT6 is used as a route-thru and the LUT5 is used as a normal LUT (I guess this is possible right?)

yes, I believe things like this can definitely occur

Reply:
FYI there are 3 ways LUTs get written:
 - Direct LUT cell placements
 - Pseudo-site pips (as wire) during site routing
 - Psuedo pips (as wire) during general routing

Reply:
I have added LUT-thru handling, and double checked with a slightly modified FF tests which yields LUT-thrus in the physical netlist.